### PR TITLE
JAMES-3028 - Split blobstore and add memory implementation

### PR DESCRIPTION
**Pull request by remk**

### CODE REVIEW

File: gc-debug-specs.adoc [hunk 1]
```diff
@@ -0,0 +1,83 @@
+= GC Debug frontend spec
```
Comment (chibenwa) [inline gc-debug-specs.adoc]: As a contributor I don't understand this document you added at the root of the project.

Maybe relocating it to a better location and better explaining what this is can help me to better understand this, or better living with ignoring it.

**Comment by Arsnael** [comment]
Ticket number on those commits please...

File: server/blob/blob-api/src/main/java/org/apache/james/blob/api/DumbBlobStore.java [hunk 2]
```diff
@@ -50,6 +50,7 @@
      * Save the blob with the provided blob id, and overwrite the previous blob with the same id if it already exists
      * The bucket is created if it not already exists.
      * This operation should be atomic and isolated
+     * Two blobs having the same blobId must have the same content
```
[the remark this answers is not in the document]
Comment (mbaechler) [reply server/blob/blob-api/src/main/java/org/apache/james/blob/api/DumbBlobStore.java]: The DumbBlobStore can't work if this assertion is not true (the technologies we use have too weak guarantees). So we ask the caller to respect this property.

File: server/blob/blob-api/src/test/java/org/apache/james/blob/api/ReadSaveDumbBlobStoreContract.java [hunk 3]
```diff
@@ -315,8 +316,8 @@ default void concurrentSaveBytesShouldReturnConsistentValues(String description,
                 (threadNumber, step) -> checkConcurrentSaveOperation(bytes)
             )
             .threadCount(10)
-            .operationCount(100)
-            .runSuccessfullyWithin(Duration.ofMinutes(2));
+            .operationCount(20)
+            .runSuccessfullyWithin(Duration.ofMinutes(10));
```
[the remark this answers is not in the document]
Comment (mbaechler) [reply server/blob/blob-api/src/test/java/org/apache/james/blob/api/ReadSaveDumbBlobStoreContract.java]: having 100 iterations make it too long. We thing that 20 iterations would trigger the bugs with a good enough probability while being 5 times shorter.
We also increased the timeout to ensure we don't have false-positive on CI (that can be slow sometimes).

File: gc-debug-specs.adoc (hunk 1 above)
Comment (remk) [reply gc-debug-specs.adoc]: I removed those two files. They don't belong to this PR indeed.

**Comment by Arsnael** [comment]
CI restart

test this please

File: server/blob/blob-api/src/main/java/org/apache/james/blob/api/DumbBlobStore.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        return save(bucketName, blobId, data.getBytes(StandardCharsets.UTF_8));
+    }
+
+    /**
+     * Remove a Blob based on its BucketName and its BlobId.
+     *
+     * @return a successful Mono if the Blob is deleted or did not exist
+     *  otherwise an IOObjectStoreException in its error channel
+     */
+    Mono<Void> delete(BucketName bucketName, BlobId blobId);
```
[the remark this answers is not in the document]
Comment (remk) [reply server/blob/blob-api/src/main/java/org/apache/james/blob/api/DumbBlobStore.java]: It should throw a null pointer exception as stated in the `DeleteDumbBlobStoreContract` : 
```
    @Test
    default void deleteShouldThrowWhenNullBucketName() {
        DumbBlobStore store = testee();
        assertThatThrownBy(() -> store.delete(null, TEST_BLOB_ID).block())
            .isInstanceOf(NullPointerException.class);
    }
```

**Comment by Arsnael** [comment]
```
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] [ERROR] Failures: 
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] [ERROR]   FixingGhostMailboxTest.ghostMailboxBugShouldDiscardOldContent:215 1 expectation failed.
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] JSON path [0][1].messageIds doesn't match.
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] Expected: a collection with size <1>
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a]   Actual: []
```

**Comment by remk** [comment]
```
eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] ]
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] [INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3.076 s - in org.apache.james.webadmin.integration.rabbitmq.RabbitMQForwardIntegrationTest
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] [INFO] 
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] [INFO] Results:
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] [INFO] 
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] [ERROR] Failures: 
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] [ERROR]   FixingGhostMailboxTest.ghostMailboxBugShouldDiscardOldContent:215 1 expectation failed.
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] JSON path [0][1].messageIds doesn't match.
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] Expected: a collection with size <1>
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a]   Actual: []
```

test this please

**Comment by Arsnael** [comment]
test this please

File: testing/base/pom.xml [hunk 5]
```diff
@@ -48,6 +48,10 @@
             <groupId>org.junit.jupiter</groupId>
             <artifactId>junit-jupiter-engine</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-params</artifactId>
```
[the remark this answers is not in the document]
Comment (Arsnael) [reply testing/base/pom.xml]: Now you can remove it logically :)

**Comment by remk** [comment]
```
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] 12:49:21.244 [ERROR] o.a.j.m.i.c.CamelProcessor - Exception calling org.apache.james.transport.mailets.Bounce: null
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] java.lang.NullPointerException
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] 	at org.apache.james.server.core.MimeMessageCopyOnWriteProxy.getWrappedMessage(MimeMessageCopyOnWriteProxy.java:140)
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] 	at org.apache.james.server.core.MimeMessageCopyOnWriteProxy.writeTo(MimeMessageCopyOnWriteProxy.java:153)
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] 	at com.sun.mail.handlers.message_rfc822.writeTo(message_rfc822.java:104)
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] 	at javax.activation.ObjectDataContentHandler.writeTo(DataHandler.java:897)
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] 	at javax.activation.DataHandler.writeTo(DataHandler.java:330)
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] 	at javax.mail.internet.MimeBodyPart.writeTo(MimeBodyPart.java:1694)
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] 	at javax.mail.internet.MimeBodyPart.writeTo(MimeBodyPart.java:996)
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] 	at javax.mail.internet.MimeMultipart.writeTo(MimeMultipart.java:561)
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] 	at com.sun.mail.handlers.multipart_mixed.writeTo(multipart_mixed.java:84)
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] 	at javax.activation.ObjectDataContentHandler.writeTo(DataHandler.java:897)
[eb2ff2c4643c18a649d611b1f3bf088aca6ba33a] 	at javax.activation.DataHandler.writeTo(DataHandler.java:330)
```

**Comment by remk** [comment]
test this please

**Comment by mbaechler** [comment]
superseded by https://github.com/linagora/james-project/pull/3085

**Pull request closed without merging**